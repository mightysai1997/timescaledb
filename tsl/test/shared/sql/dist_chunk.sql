-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

-- This file contains tests for all features that will be used as part
-- of the chunk move/copy multi-node functionality


-- A table for the first chunk will be created on the data node, where it is not present.
SELECT chunk_name, data_nodes
FROM timescaledb_information.chunks
WHERE hypertable_name = 'dist_chunk_copy'
ORDER BY 1, 2;

-- Non-distributed chunk will be used to test an error
SELECT chunk_name
FROM timescaledb_information.chunks
WHERE hypertable_name = 'conditions'
ORDER BY 1;

-- Test function _timescaledb_functions.create_chunk_replica_table
\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.create_chunk_replica_table(NULL, 'data_node_1');
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', NULL);
SELECT _timescaledb_functions.create_chunk_replica_table(1234, 'data_node_1');
SELECT _timescaledb_functions.create_chunk_replica_table('metrics_int', 'data_node_1');
SELECT _timescaledb_functions.create_chunk_replica_table('conditions', 'data_node_1');
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._hyper_10_51_chunk', 'data_node_1');
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', 'data_node_2');
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_27_chunk', 'data_node_2');
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', 'data_node_4');
BEGIN READ ONLY;
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', 'data_node_3');
COMMIT;
\set ON_ERROR_STOP 1

SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', 'data_node_1');

-- Test that the table cannot be created since it was already created on the data node
\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_67_chunk', 'data_node_3');
\set ON_ERROR_STOP 1

-- Creating chunk replica table ignores compression now:
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_15_70_chunk', 'data_node_2');

CALL distributed_exec($$ DROP TABLE _timescaledb_internal._dist_hyper_15_67_chunk $$, '{"data_node_1"}');
CALL distributed_exec($$ DROP TABLE _timescaledb_internal._dist_hyper_15_70_chunk $$, '{"data_node_2"}');

-- Test function _timescaledb_functions.chunk_drop_replica
-- Sanity checking of the chunk_drop_replica API

\set ON_ERROR_STOP 0
-- Check that it doesn't work in a read only transaction
SET default_transaction_read_only TO on;
SELECT _timescaledb_functions.chunk_drop_replica(NULL, NULL);
RESET default_transaction_read_only;

-- NULL input for chunk id errors out
SELECT _timescaledb_functions.chunk_drop_replica(NULL, NULL);

-- Specifying any regular hypertable instead of chunk errors out
SELECT _timescaledb_functions.chunk_drop_replica('public.metrics', NULL);

-- Specifying regular hypertable chunk on a proper data node errors out
SELECT _timescaledb_functions.chunk_drop_replica('_timescaledb_internal._hyper_1_1_chunk', 'data_node_1');

-- Specifying non-existent chunk on a proper data node errors out
SELECT _timescaledb_functions.chunk_drop_replica('_timescaledb_internal._dist_hyper_700_38_chunk', 'data_node_1');

-- Get the last chunk for this hypertable
SELECT ch1.schema_name|| '.' || ch1.table_name as "CHUNK_NAME", ch1.id "CHUNK_ID"
FROM _timescaledb_catalog.chunk ch1, _timescaledb_catalog.hypertable ht
WHERE ch1.hypertable_id = ht.id
AND ht.table_name = 'mvcp_hyper'
ORDER BY ch1.id DESC LIMIT 1 \gset

-- Specifying wrong node name errors out
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', 'bad_node');

-- This chunk contains only one entry as of now
SELECT * FROM :CHUNK_NAME;

-- Specifying NULL node name along with proper chunk errors out
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', NULL);

\set ON_ERROR_STOP 1
-- Check the current primary foreign server for this chunk, that will change
-- post the chunk_drop_replica call
SELECT foreign_server_name AS "PRIMARY_CHUNK_NODE"
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1 \gset

-- Show the node that was primary for the chunk
\echo :PRIMARY_CHUNK_NODE

-- Drop the chunk replica on the primary chunk node. Should succeed
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', :'PRIMARY_CHUNK_NODE');

-- The primary foreign server for the chunk should be updated now
SELECT foreign_server_name
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1;

-- Number of replicas should have been reduced by 1
SELECT count(*) FROM _timescaledb_catalog.chunk_data_node WHERE chunk_id = :'CHUNK_ID';

-- Ensure that INSERTs still work on this mvcp_hyper table into this chunk
-- Rollback to not modify the shared test state
BEGIN;
INSERT INTO mvcp_hyper VALUES (1001, 1001);
-- Ensure that SELECTs are able to query data from the above chunk
SELECT count(*) FROM mvcp_hyper WHERE time >= 1000;
ROLLBACK;


-- Check that chunk_drop_replica works with compressed chunk
SELECT substr(compress_chunk(:'CHUNK_NAME')::TEXT, 1, 29);

-- Drop one replica of a valid chunk. Should succeed on another datanode
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', 'data_node_2');

-- Number of replicas should have been reduced by 1
SELECT count(*) FROM _timescaledb_catalog.chunk_data_node WHERE chunk_id = :'CHUNK_ID';

-- Decompress before checking INSERTs
SELECT substr(decompress_chunk(:'CHUNK_NAME')::TEXT, 1, 29);

-- Ensure that INSERTs still work on this mvcp_hyper table into this chunk
-- Rollback to not modify the shared test state
BEGIN;
INSERT INTO mvcp_hyper VALUES (1002, 1002);
-- Ensure that SELECTs are able to query data from the above chunk
SELECT count(*) FROM mvcp_hyper WHERE time >= 1000;
ROLLBACK;

-- Drop one replica of a valid chunk. Should not succeed on last datanode
SELECT foreign_server_name AS "PRIMARY_CHUNK_NODE"
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1 \gset

\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', :'PRIMARY_CHUNK_NODE');
\set ON_ERROR_STOP 1
