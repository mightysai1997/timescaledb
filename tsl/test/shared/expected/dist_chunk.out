-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
-- This file contains tests for all features that will be used as part
-- of the chunk move/copy multi-node functionality
-- A table for the first chunk will be created on the data node, where it is not present.
SELECT chunk_name, data_nodes
FROM timescaledb_information.chunks
WHERE hypertable_name = 'dist_chunk_copy'
ORDER BY 1, 2;
       chunk_name        |        data_nodes         
-------------------------+---------------------------
 _dist_hyper_X_X_chunk | {data_node_2,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_2,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_2,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_2}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_2,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_2}
 _dist_hyper_X_X_chunk | {data_node_2,data_node_3}
 _dist_hyper_X_X_chunk | {data_node_1,data_node_3}
(11 rows)

-- Non-distributed chunk will be used to test an error
SELECT chunk_name
FROM timescaledb_information.chunks
WHERE hypertable_name = 'conditions'
ORDER BY 1;
     chunk_name     
 _hyper_X_X_chunk
 _hyper_X_X_chunk
(2 rows)

-- Test function _timescaledb_functions.create_chunk_replica_table
\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.create_chunk_replica_table(NULL, 'data_node_1');
ERROR:  chunk cannot be NULL
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', NULL);
ERROR:  data node name cannot be NULL
SELECT _timescaledb_functions.create_chunk_replica_table(1234, 'data_node_1');
ERROR:  oid "1234" is not a chunk
SELECT _timescaledb_functions.create_chunk_replica_table('metrics_int', 'data_node_1');
ERROR:  relation "metrics_int" is not a chunk
SELECT _timescaledb_functions.create_chunk_replica_table('conditions', 'data_node_1');
ERROR:  relation "conditions" is not a chunk
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._hyper_X_X_chunk', 'data_node_1');
ERROR:  chunk "_hyper_X_X_chunk" doesn't belong to a distributed hypertable
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_2');
ERROR:  chunk "_dist_hyper_X_X_chunk" already exists on data node "data_node_2"
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_2');
ERROR:  relation "_timescaledb_internal._dist_hyper_X_X_chunk" does not exist at character 58
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_4');
ERROR:  server "data_node_4" does not exist
BEGIN READ ONLY;
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_3');
ERROR:  cannot execute create_chunk_replica_table() in a read-only transaction
COMMIT;
\set ON_ERROR_STOP 1
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_1');
 create_chunk_replica_table 
 
(1 row)

-- Test that the table cannot be created since it was already created on the data node
\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_3');
ERROR:  chunk "_dist_hyper_X_X_chunk" already exists on data node "data_node_3"
\set ON_ERROR_STOP 1
-- Creating chunk replica table ignores compression now:
SELECT _timescaledb_functions.create_chunk_replica_table('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_2');
 create_chunk_replica_table 
 
(1 row)

CALL distributed_exec($$ DROP TABLE _timescaledb_internal._dist_hyper_X_X_chunk $$, '{"data_node_1"}');
CALL distributed_exec($$ DROP TABLE _timescaledb_internal._dist_hyper_X_X_chunk $$, '{"data_node_2"}');
-- Test function _timescaledb_functions.chunk_drop_replica
-- Sanity checking of the chunk_drop_replica API
\set ON_ERROR_STOP 0
-- Check that it doesn't work in a read only transaction
SET default_transaction_read_only TO on;
SELECT _timescaledb_functions.chunk_drop_replica(NULL, NULL);
ERROR:  cannot execute chunk_drop_replica() in a read-only transaction
RESET default_transaction_read_only;
-- NULL input for chunk id errors out
SELECT _timescaledb_functions.chunk_drop_replica(NULL, NULL);
ERROR:  invalid chunk relation
-- Specifying any regular hypertable instead of chunk errors out
SELECT _timescaledb_functions.chunk_drop_replica('public.metrics', NULL);
ERROR:  invalid chunk relation
-- Specifying regular hypertable chunk on a proper data node errors out
SELECT _timescaledb_functions.chunk_drop_replica('_timescaledb_internal._hyper_X_X_chunk', 'data_node_1');
ERROR:  "_hyper_X_X_chunk" is not a valid remote chunk
-- Specifying non-existent chunk on a proper data node errors out
SELECT _timescaledb_functions.chunk_drop_replica('_timescaledb_internal._dist_hyper_X_X_chunk', 'data_node_1');
ERROR:  relation "_timescaledb_internal._dist_hyper_X_X_chunk" does not exist at character 50
-- Get the last chunk for this hypertable
SELECT ch1.schema_name|| '.' || ch1.table_name as "CHUNK_NAME", ch1.id "CHUNK_ID"
FROM _timescaledb_catalog.chunk ch1, _timescaledb_catalog.hypertable ht
WHERE ch1.hypertable_id = ht.id
AND ht.table_name = 'mvcp_hyper'
ORDER BY ch1.id DESC LIMIT 1 \gset
-- Specifying wrong node name errors out
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', 'bad_node');
ERROR:  server "bad_node" does not exist
-- This chunk contains only one entry as of now
SELECT * FROM :CHUNK_NAME;
 time | value 
------+-------
 1000 |  1000
(1 row)

-- Specifying NULL node name along with proper chunk errors out
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', NULL);
ERROR:  data node name cannot be NULL
\set ON_ERROR_STOP 1
-- Check the current primary foreign server for this chunk, that will change
-- post the chunk_drop_replica call
SELECT foreign_server_name AS "PRIMARY_CHUNK_NODE"
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1 \gset
-- Show the node that was primary for the chunk
\echo :PRIMARY_CHUNK_NODE
data_node_1
-- Drop the chunk replica on the primary chunk node. Should succeed
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', :'PRIMARY_CHUNK_NODE');
 chunk_drop_replica 
 
(1 row)

-- The primary foreign server for the chunk should be updated now
SELECT foreign_server_name
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1;
 foreign_server_name 
 data_node_2
(1 row)

-- Number of replicas should have been reduced by 1
SELECT count(*) FROM _timescaledb_catalog.chunk_data_node WHERE chunk_id = :'CHUNK_ID';
 count 
     2
(1 row)

-- Ensure that INSERTs still work on this mvcp_hyper table into this chunk
-- Rollback to not modify the shared test state
BEGIN;
INSERT INTO mvcp_hyper VALUES (1001, 1001);
-- Ensure that SELECTs are able to query data from the above chunk
SELECT count(*) FROM mvcp_hyper WHERE time >= 1000;
 count 
     2
(1 row)

ROLLBACK;
-- Check that chunk_drop_replica works with compressed chunk
SELECT substr(compress_chunk(:'CHUNK_NAME')::TEXT, 1, 29);
            substr             
 _timescaledb_internal._dist_h
(1 row)

-- Drop one replica of a valid chunk. Should succeed on another datanode
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', 'data_node_2');
 chunk_drop_replica 
 
(1 row)

-- Number of replicas should have been reduced by 1
SELECT count(*) FROM _timescaledb_catalog.chunk_data_node WHERE chunk_id = :'CHUNK_ID';
 count 
     1
(1 row)

-- Decompress before checking INSERTs
SELECT substr(decompress_chunk(:'CHUNK_NAME')::TEXT, 1, 29);
            substr             
 _timescaledb_internal._dist_h
(1 row)

-- Ensure that INSERTs still work on this mvcp_hyper table into this chunk
-- Rollback to not modify the shared test state
BEGIN;
INSERT INTO mvcp_hyper VALUES (1002, 1002);
-- Ensure that SELECTs are able to query data from the above chunk
SELECT count(*) FROM mvcp_hyper WHERE time >= 1000;
 count 
     2
(1 row)

ROLLBACK;
-- Drop one replica of a valid chunk. Should not succeed on last datanode
SELECT foreign_server_name AS "PRIMARY_CHUNK_NODE"
    FROM information_schema.foreign_tables WHERE
    foreign_table_name = split_part(:'CHUNK_NAME', '.', 2)
    ORDER BY 1 \gset
\set ON_ERROR_STOP 0
SELECT _timescaledb_functions.chunk_drop_replica(:'CHUNK_NAME', :'PRIMARY_CHUNK_NODE');
ERROR:  cannot drop the last chunk replica
\set ON_ERROR_STOP 1
